src/rdydata.c

```diff
@@ -89,54 +110,274 @@ PetscErrorCode RDySetDirichletBoundaryValues(RDy rdy, PetscInt boundary_index, P
   PetscFunctionReturn(PETSC_SUCCESS);
 }
 
-PetscErrorCode RDyGetHeight(RDy rdy, PetscReal *h) {
+static PetscErrorCode RDyGetPrognosticVariableOfLocalCell(RDy rdy, PetscInt idof, PetscReal *values) {
   PetscFunctionBegin;
 
   PetscReal *x;
   PetscCall(VecGetArray(rdy->X, &x));
   for (PetscInt i = 0; i < rdy->mesh.num_cells_local; ++i) {
-    h[i] = x[3 * i];
+    values[i] = x[3 * i + idof];
   }
   PetscCall(VecRestoreArray(rdy->X, &x));
   PetscFunctionReturn(PETSC_SUCCESS);
 }
 
-PetscErrorCode RDyGetXVelocity(RDy rdy, PetscReal *vx) {
+PetscErrorCode RDyGetHeightOfLocalCell(RDy rdy, const PetscInt size, PetscReal values[size]) {
+  PetscFunctionBegin;
+  PetscCall(CheckNumLocalCells(rdy, size));
+  PetscInt idof = 0;
+  PetscCall(RDyGetPrognosticVariableOfLocalCell(rdy, idof, values));
+  PetscFunctionReturn(PETSC_SUCCESS);
+}
+
+PetscErrorCode RDyGetXMomentumOfLocalCell(RDy rdy, const PetscInt size, PetscReal values[size]) {
   PetscFunctionBegin;
+  PetscCall(CheckNumLocalCells(rdy, size));
+  PetscInt idof = 1;
+  PetscCall(RDyGetPrognosticVariableOfLocalCell(rdy, idof, values));
+  PetscFunctionReturn(PETSC_SUCCESS);
+}
 
-  PetscReal *x;
-  PetscCall(VecGetArray(rdy->X, &x));
+PetscErrorCode RDyGetYMomentumOfLocalCell(RDy rdy, const PetscInt size, PetscReal values[size]) {
+  PetscFunctionBegin;
+  PetscCall(CheckNumLocalCells(rdy, size));
+  PetscInt idof = 2;
+  PetscCall(RDyGetPrognosticVariableOfLocalCell(rdy, idof, values));
+  PetscFunctionReturn(PETSC_SUCCESS);
+}
+
+PetscErrorCode RDySetSourceVecForLocalCell(RDy rdy, Vec src_vec, PetscInt idof, PetscReal *values) {
+  PetscFunctionBegin;
+
+  PetscInt ndof;
+  PetscCall(VecGetBlockSize(src_vec, &ndof));
+
+  PetscAssert(idof < ndof, PETSC_COMM_WORLD, PETSC_ERR_ARG_SIZ, "Trying to set values for a block that exceeds total number of blocks");
```

Comment:
It may be easier for folks to understand `idof` as a "component of the solution vector", which in this case is invalid. In any case, the associated error here is not that `idof` refers to a block in the associated operator matrix, but that `idof` is an index within a block, and this index could exceed the block size.